### PR TITLE
[RFC] vim-patch: 8.0.0251 (pyxversion, :pythonx)

### PR DESCRIPTION
**Pull request by dvejmz**
Problem: It is not so easy to write a script that works with both Python 2 and Python 3, even when the Python code works with both.
Solution: Add 'pyxversion', :pyx, etc. (Marc Weber, Ken Takata)

https://github.com/vim/vim/commit/f42dd3c3901ea0ba38e67a616aea9953cae81b8d

This patch had a lot of feature-switching logic, most of it irrelevant to Neovim, so I've removed it whilst retaining the intended behaviour as best as I could. Areas that have been left out are mostly Python feature `#define`s, Ex commands and non-existent files like `Filelist` and test Makefile (`testdir/Make_all.mak`)

Also, very minor refactors on the ported code (curly braces, etc.).

Still a WIP as I plan to do some more work on it (as well as address any feedback I may get if anything stands out to anyone already). I will squash all commits into the template one before the RFC stage.

### CODE REVIEW

**Comment by justinmk** [comment]
Would be great to get this in. What's remaining?

**Comment by bfredl** [comment]
python related `#ifdef`:s has to also be removed. Instead of compile-time checks and `python3_enabled`, `eval_has_provider` should be used (with some care, as the later is potentially a more expensive operation).

**Comment by dvejmz** [comment]
My apologies, I've been unable to work on this for the last few day. I'll have a look at your comments now and address them.

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -4452,6 +4452,30 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+		only |+python3|			3
+
+	Available values are 0, 2 and 3.
+	If 'pyxversion' is 0, it is set to 2 or 3 after the first execution of
+	any python2/3 commands or functions.  E.g. `:py` sets to 2, and `:py3`
+	sets to 3. `:pyx` sets it to 3 if Python 3 is available, otherwise sets
+	to 2 if Python 2 is available.
+	See also: |has-pythonx|
+
+	If Vim is compiled with only |+python| or |+python3| setting
```
[the remark this answers is not in the document]
Comment (dvejmz) [reply runtime/doc/options.txt]: Changed

**Comment by dvejmz** [comment]
Updated PR as per feedback.

**Comment by dvejmz** [comment]
Fixed @bfredl , much simpler now, no doubt!

**Comment by bfredl** [comment]
Please also check the travis lint build for some remaining issues (latest build https://travis-ci.org/neovim/neovim/jobs/454240981). In particular use two spaces, never tabs, and `//` comments only.

File: src/nvim/option.c [hunk 2]
```diff
@@ -263,6 +263,9 @@ typedef struct vimoption {
   "!:CursorColumn,.:CursorLine,o:ColorColumn,q:QuickFixLine," \
   "0:Whitespace,I:NormalNC"
 
+// Default python version for pyx* commands
+# define DEFAULT_PYTHON_VER 0
```
Comment (justinmk) [inline src/nvim/option.c]: this isn't used

File: src/nvim/option.c [hunk 3]
```diff
@@ -4329,6 +4332,10 @@ static char *set_num_option(int opt_idx, char_u *varp, long value,
     if (p_uc && !old_value) {
       ml_open_files();
     }
+  } else if (pp == &p_pyx) {
+    if (p_pyx != 0 && p_pyx != 2 && p_pyx != 3) {
+        errmsg = e_invarg;
```
Comment (justinmk) [inline src/nvim/option.c]: 2-space indent

**Comment by justinmk** [comment]
CI is reporting an error (please check the CI logs, if you're wondering why the build is red):

```
../src/nvim/eval.c:13058:13: error: unused function 'f_pyxeval' [-Werror,-Wunused-function]
static void f_pyxeval(typval_T *argvars, typval_T *rettv)
            ^
1 error generated.
```

File: src/nvim/eval.c [hunk 4]
```diff
@@ -13042,9 +13047,24 @@ static void f_pyeval(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
 {
+  if (p_pyx == 0) {
+      p_pyx = 3;
+  }
+
   script_host_eval("python3", argvars, rettv);
 }
 
+// "pyxeval()" function
+static void f_pyxeval(typval_T *argvars, typval_T *rettv)
```
Comment (justinmk) [inline src/nvim/eval.c]: This is missing the `FunPtr fptr` parameter, which is why the build is broken...

File: runtime/doc/options.txt [hunk 5]
```diff
@@ -4451,6 +4451,30 @@ A jump table for the options with a short description can be found at |Q_op|.
 	Insert mode completion.  When zero as much space as available is used.
 	|ins-completion-menu|.
 
+						*'pyxversion'* *'pyx'*
+'pyxversion' 'pyx'	number	(default depends on the build)
+			global
+			{not in Vi}
+			{only available when compiled with the |+python| or
+			the |+python3| feature}
```
Comment (justinmk) [inline runtime/doc/options.txt]: can remove this "compiled with" note

File: runtime/doc/options.txt [hunk 6]
```diff
@@ -4451,6 +4451,30 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+	Specifies the python version used for pyx* functions and commands
+	|python_x|.  The default value is as follows:
+
+		Compiled with		     Default ~
```
Comment (justinmk) [inline runtime/doc/options.txt]: instead of `Compiled with`, how about `|provider| installed`

**Comment by dvejmz** [comment]
Sure, I was planning to squash everything in and add the proper commit message once the I had approval on the PR. I'll have a look at this this week.

**Comment by dvejmz** [comment]
I've addressed your latest comments minus this

> This is missing `ex_python3` and the associated change to `ex_cmds.lua`. So `:pyx` command is not present.

I can port over this change too although you mentioned `ex_python3` is missing. I do not understand what you mean by this as [the original PR commit](https://github.com/vim/vim/commit/f42dd3c3901ea0ba38e67a616aea9953cae81b8d) doesn't include any such definition. It seems to be already defined in `src/nvim/ex_cmds.c` though

```
void ex_python3(exarg_T *eap)
{
  script_host_execute("python3", eap);
}
```

If this is not what you meant, could you be referring to some other required definition of `ex_python3` I don't know about? I am sorry but my knowledge of Neovim is very limited!

**Comment by dvejmz** [comment]
QuickBuild is now failing due to `f_pyxeval` not being unused, which, on a local build is a compiler WARN, but is flagged up as an error on QuickBuild. The function is not used anywhere in the original PR either so I'm unsure what I should do. Should I remove it?

**Comment by bfredl** [comment]
> Do functions defined in /src/nvim/eval.c need to be declared somewhere else?

Yes, in `src/nvim/eval.lua`. See the existing definition of `py3eval`.

**Comment by justinmk** [comment]
Pushed an update with some fixes.

**Pull request merged**